### PR TITLE
feat: enables to run ike against particular namespace

### LINKED ISSUE #103: Ability to specify a namespace we want to develop against
Currently, we rely on default namespace being defined in kube context. This, however, limits our testing possibilities, as we might want to use `ike` for two different namespaces at the same time.

Since telepresence allows to specify `--namespace` as an argument we should add corresponding flag and delegate.

### PR DESCRIPTION
**Pull request by bartoszmajsak**
Fixes #103

### CODE REVIEW

**Comment by aslakknutsen** [comment]
Is this not missing the namespace being used by the session_client? TP would run against different namespaces, but the session would never be made in the correct namespace so TP wouldn't find the D/DC?

**Comment by bartoszmajsak** [comment]
> Is this not missing the namespace being used by the session_client? TP would run against different namespaces, but the session would never be made in the correct namespace so TP wouldn't find the D/DC?

You are totally right... not sure how I missed that

**Comment by bartoszmajsak** [comment]
@aslakknutsen fixed with de8a5f7

File: cmd/ike/session/session.go [hunk 1]
```diff
@@ -143,26 +135,26 @@ func waitForRefToComplete(sessionName, ref string) (string, error) {
...
+		_ = h.c.Delete(session)
 	} else {
-		_ = DefaultClient().Create(session)
+		_ = h.c.Create(session)
 	}
 }
 
 func getSessionName(cmd *cobra.Command) string {
-	sessionName, err := cmd.Flags().GetString("session")
+	sessionName, err := cmd.Flags().GetString("sessionName")
```
Comment (aslakknutsen) [inline cmd/ike/session/session.go]: That this passes our test suite is a little scary: https://github.com/aslakknutsen/istio-workspace/blob/de8a5f7c4c09e8e98349ab1d59e72b52ec3e9427/cmd/ike/cmd/develop.go#L79

File: cmd/ike/session/session.go [hunk 2]
```diff
@@ -143,26 +135,26 @@ func waitForRefToComplete(sessionName, ref string) (string, error) {
...
 
-func removeOrLeaveSession(sessionName, ref string) {
-	session, err := DefaultClient().Get(sessionName)
+func (h *handler) removeOrLeaveSession() {
+	session, err := h.c.Get(h.sessionName)
 	if err != nil {
 		return // assume missing, nothing to clean?
 	}
-	// more then one participant, update session
+	// more then one participant, update sessionName
```
Comment (aslakknutsen) [inline cmd/ike/session/session.go]: search-replace to wide. makes no sense.

File: cmd/ike/session/session.go [hunk 3]
```diff
@@ -43,32 +59,32 @@ func CreateOrJoinHandler(cmd *cobra.Command) (func(), error) {
...
 			return "", err
 		}
-		return waitForRefToComplete(sessionName, ref)
+		return h.waitForRefToComplete()
 	}
-	// join session
-
-	session.Spec.Refs = append(session.Spec.Refs, ref)
-	err = DefaultClient().Create(session)
+	// join sessionName
```
Comment (aslakknutsen) [inline cmd/ike/session/session.go]: search-replace to wide. makes no sense.

File: cmd/ike/session/session.go [hunk 4]
```diff
@@ -15,25 +15,41 @@ import (
...
+	sessionName,
+	route       string
+}
+
+// CreateOrJoinHandler provides the option to either create a new sessionName if non exist or join an existing.
 // Rely on the following flags:
+//  * namespace - the name of the target namespace where deployment is defined
 //  * deployment - the name of the target deployment and will update the flag with the new deployment name
-//  * session - the name of the session
+//  * sessionName - the name of the sessionName
```
Comment (aslakknutsen) [inline cmd/ike/session/session.go]: search-replace to wide. makes no sense.

File: cmd/ike/session/session.go [hunk 5]
```diff
@@ -15,25 +15,41 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 )
 
-// Handler is a function to setup a server session before attempting to connect. Returns a 'cleanup' function.
+// Handler is a function to setup a server sessionName before attempting to connect. Returns a 'cleanup' function.
```
Comment (aslakknutsen) [inline cmd/ike/session/session.go]: search-replace to wide. makes no sense.

File: cmd/ike/session/session_client_test.go [hunk 6]
```diff
@@ -20,7 +20,7 @@ var _ = Describe("Session Client operations", func() {
 			Kind:       "Session",
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			Name: "sample-session",
+			Name: "sample-sessionName",
```
Comment (aslakknutsen) [inline cmd/ike/session/session_client_test.go]: search-replace to wide. makes no sense.

File: cmd/ike/session/session_client_test.go [hunk 7]
```diff
@@ -34,11 +34,11 @@ var _ = Describe("Session Client operations", func() {
 		fakeClient := testclient.NewSimpleClientset()
 		client, _ := session.NewClient(fakeClient, "test-namespace")
 
-		It("should get created session byt its name", func() {
+		It("should get created sessionName byt its name", func() {
```
Comment (aslakknutsen) [inline cmd/ike/session/session_client_test.go]: search-replace to wide. makes no sense.

File: cmd/ike/session/session_client_test.go [hunk 8]
```diff
@@ -56,11 +56,11 @@ var _ = Describe("Session Client operations", func() {
 			Expect(creationErr).ToNot(HaveOccurred())
 		})
 
-		It("should delete session byt its name", func() {
+		It("should delete sessionName byt its name", func() {
```
Comment (aslakknutsen) [inline cmd/ike/session/session_client_test.go]: search-replace to wide. makes no sense.

File: cmd/ike/session/session.go (hunk 1 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session.go]: Lemme have a look :) but yeah, it is

File: cmd/ike/session/session.go (hunk 3 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session.go]:
```suggestion
	// join session
```

File: cmd/ike/session/session.go (hunk 5 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session.go]:
```suggestion
// Handler is a function to setup a server sessionName before attempting to connect. Returns a 'cleanup' function.
```

File: cmd/ike/session/session_client_test.go (hunk 7 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session_client_test.go]:
```suggestion
It("should get created sessionName byt its name"
```

File: cmd/ike/session/session.go (hunk 1 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session.go]:
```suggestion
	sessionName, err := cmd.Flags().GetString("session")
```

File: cmd/ike/session/session.go (hunk 4 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session.go]:
```suggestion
//  * session - the name of the session
```

File: cmd/ike/session/session_client_test.go (hunk 8 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session_client_test.go]:
```suggestion
		It("should delete session by its name", func() {
```

File: cmd/ike/session/session_client_test.go (hunk 6 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session_client_test.go]:
```suggestion
			Name: "sample-session",
```

File: cmd/ike/session/session.go (hunk 2 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session.go]:
```suggestion
	// more than one participant, update session
```

File: cmd/ike/session/session.go (hunk 1 above)
Comment (bartoszmajsak) [reply cmd/ike/session/session.go]: It seems we don't cover a case when we want to join an existing session - so even with this regression we anyway always created a new one. 

Joining existing would be a bit tricky with starting 2 TP instances, but I can provide a test where I create Session CR upfront and join it.

Comment (aslakknutsen) [reply cmd/ike/session/session.go]: Hmm, shouldn't we technically end in a "err" here? https://github.com/aslakknutsen/istio-workspace/blob/36cdc4daddd549aa9577dac27cc253d77de8dc24/cmd/ike/session/session.go#L158 

Or is that only on type parse error?

Comment (bartoszmajsak) [reply cmd/ike/session/session.go]: Yeah, that too. I already changed that.

About joining an existing session - maybe we could provide an option to persist a session, as currently when we stop `ike develop` we always clean up. How about opening an issue for that and ship the test together with it?

Comment (aslakknutsen) [reply cmd/ike/session/session.go]: Hmm.. That would leave a Route in error when we tear down the Local service it points to. We could for the sake of test, or in prep for a future feature of 'the monitoring option/build image' replacement type of thing... 

Maybe just leave it as to why we need this? https://github.com/aslakknutsen/istio-workspace/issues/59 :)

Comment (bartoszmajsak) [reply cmd/ike/session/session.go]: > That would leave a Route in error when we tear down the Local service it points to.

Correct - we would need to check if there are any "active" participants first, before routing. This also opens up a question on how to handle two participants working on the same service anyway (from the session concept point of view). Something we should discuss in depth I believe

Comment (bartoszmajsak) [reply cmd/ike/session/session.go]: I added logging, as this is parse error most likely. In such a case, I think we should roll back to random session name.

File: cmd/ike/session/session_client.go [hunk 9]
```diff
@@ -1,7 +1,7 @@
 package session
 
 import (
-	"log"
+	logger "log"
```
Comment (aslakknutsen) [inline cmd/ike/session/session_client.go]: Different logger?

Comment (bartoszmajsak) [reply cmd/ike/session/session_client.go]: One is on the client side and the other within the cluster. We can use the same though, won't harm

Comment (bartoszmajsak) [reply cmd/ike/session/session_client.go]: Or is it...

Comment (bartoszmajsak) [reply cmd/ike/session/session_client.go]: Fixed with #fd2c88f

Comment (bartoszmajsak) [reply cmd/ike/session/session_client.go]: and bb11168

**Pull request merged**